crypto/Pos/Crypto/Orphans.hs

```diff
@@ -8,51 +8,149 @@ import           Universum
 import           Crypto.Scrypt (EncryptedPass (..))
-import qualified Crypto.Sign.Ed25519 as Ed25519
 import           Data.Aeson (FromJSON (..), ToJSON (..))
+import           Data.Aeson.Types (Parser)
+import qualified Data.ByteArray as BA
+import qualified Data.ByteString as BS
 import           Data.Hashable (Hashable)
-import           Data.SafeCopy (base, deriveSafeCopySimple)
+import           Data.SafeCopy (SafeCopy (..), base, contain,
+                     deriveSafeCopySimple, safeGet, safePut)
+import qualified Data.Text as T
 import           Serokell.Util.Base64 (JsonByteString (..))
 
 import           Pos.Binary.Class (Bi (..), decodeBinary, encodeBinary)
 
-instance Hashable Ed25519.PublicKey
-instance Hashable Ed25519.SecretKey
-instance Hashable Ed25519.Signature
+instance Hashable Ed25519.PublicKey where
+  hashWithSalt salt pk = hashWithSalt salt $ (BA.convert pk :: BS.ByteString)
+
+instance Hashable Ed25519.SecretKey where
+  hashWithSalt salt pk = hashWithSalt salt $ (BA.convert pk :: BS.ByteString)
+
+instance Hashable Ed25519.Signature where
+  hashWithSalt salt pk = hashWithSalt salt $ (BA.convert pk :: BS.ByteString)
+
+
+instance Ord Ed25519.PublicKey where
+  compare x1 x2 = compare (BA.convert x1 :: BS.ByteString) (BA.convert x2 :: BS.ByteString)
+
+instance Ord Ed25519.SecretKey where
+  compare x1 x2 = compare (BA.convert x1 :: BS.ByteString) (BA.convert x2 :: BS.ByteString)
+
+instance Ord Ed25519.Signature where
+  compare x1 x2 = compare (BA.convert x1 :: BS.ByteString) (BA.convert x2 :: BS.ByteString)
+
+
+instance SafeCopy BA.Bytes where
+  putCopy s =
+    let
+      toByteString :: BA.Bytes -> BS.ByteString
+      toByteString = BA.convert
+    in
+      contain $ safePut (toByteString s)
+  getCopy =
+    let
+      fromByteString :: BS.ByteString -> BA.Bytes
+      fromByteString = BA.convert
+    in
+      contain $ fromByteString <$> safeGet
+
+
+instance SafeCopy BA.ScrubbedBytes where
+  putCopy s =
+    let
+      toByteString :: BA.ScrubbedBytes -> BS.ByteString
+      toByteString = BA.convert
+    in
+      contain $ safePut (toByteString s)
+  getCopy =
+    let
+      fromByteString :: BS.ByteString -> BA.ScrubbedBytes
+      fromByteString = BA.convert
+    in
+      contain $ fromByteString <$> safeGet
 
-instance NFData Ed25519.PublicKey
-instance NFData Ed25519.SecretKey
-instance NFData Ed25519.Signature
 
 deriveSafeCopySimple 0 'base ''Ed25519.PublicKey
 deriveSafeCopySimple 0 'base ''Ed25519.SecretKey
 deriveSafeCopySimple 0 'base ''Ed25519.Signature
 
+fromByteString :: BS.ByteString -> BA.Bytes
+fromByteString = BA.convert
+
 instance FromJSON Ed25519.PublicKey where
-    parseJSON v = Ed25519.PublicKey . getJsonByteString <$> parseJSON v
+  parseJSON v =
+    let
+      fromCryptoToAeson :: CryptoFailable a -> Parser a
+      fromCryptoToAeson (CryptoFailed e) = error $ mappend "Pos.Crypto.Orphan.parseJSON Ed25519.PublicKey failed because " (T.pack $ show e)
```

Comment:
Instead of `error`, use `fail` -- `fail :: MonadFail m => String -> m a` allows the parser monad to report failure, while `error :: String -> a` causes a value that will throw an exception when forced, bypassing most error handling mechanisms.